network_simulator/link.py

```diff
@@ -4,44 +4,98 @@
 """
 
 from Queue import Queue
+from packet import *
+from common import *
+from event import *
+import logging
+
+logger = logging.getLogger(__name__)
 
-# TODO(yubo): Finish implementing the classes and Events for this.
 
 
 class LinkBuffer(object):
     """
     Representation of a Link's FIFO buffer.
-
-    TODO(team): Parameters related to link congestion, so we can have dynamic
-    routing table updates.
-    TODO(yubo): Functions for whether this is full etc.
     """
 
-    def __init__(self, buffer_size_bits=None, queue=Queue()):
+    # TODO(team): Parameters related to link congestion, so we can have dynamic
+    # routing table updates.
+    # TODO(yubo): packet_exemption_times deque
+
+    def __init__(self, max_buffer_size_bits=None, queue=Queue()):
         """
-        :ivar int buffer_size_bits: maximum buffer size in bits.
+        :ivar int max_buffer_size_bits: maximum buffer size in bits.
         :ivar Queue queue: a FIFO queue of LinkBufferElements
+        :ivar int curr_buffer_size_bits: current buffer size in bits
         """
-        self.buffer_size_bits = buffer_size_bits
+        self.max_buffer_size_bits = max_buffer_size_bits
         self.queue = queue
 
     def __repr__(self):
         return str(self.__dict__)
 
+    def remaining(self):
```

Comment:
Maybe rename to "remaining_size_bits". This can be done by selecting "remaining" and pressing Shift+F6 in PyCharm.